zipkin/src/main/java/zipkin2/internal/ByteArrayBuffer.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2015-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.internal;
+
+final class ByteArrayBuffer extends Buffer {
+
+  private final byte[] buf;
+  int pos; // visible for testing
+
+  ByteArrayBuffer(int size) {
+    buf = new byte[size];
+  }
+
+  ByteArrayBuffer(byte[] buf, int pos) {
+    this.buf = buf;
+    this.pos = pos;
+  }
+
+  @Override public void writeByte(int v) {
+    buf[pos++] = (byte) (v & 0xff);
+  }
+
+  @Override public void write(byte[] v) {
+    System.arraycopy(v, 0, buf, pos, v.length);
+    pos += v.length;
+  }
+
+  @Override void writeBackwards(long v) {
```

Comment:
this is special cased as I can't think a better way to write a long as ascii.. and don't really want to think I know how a given buffer type will write a value backwards. For example, if using okio.buffer we don't need to implement this at all as it already has a native function to write a decimal as ascii..